### PR TITLE
Enable deleting of currently downloading map & add option to stop download from notification

### PR DESCRIPTION
**Pull request by ekigamba**

### CODE REVIEW

File: library/src/main/java/io/ona/kujaku/data/realm/RealmDatabase.java [hunk 1]
```diff
@@ -33,4 +35,33 @@ private RealmDatabase(Context context) {
                 .build();
         Realm.setDefaultConfiguration(configuration);
     }
+
+    public boolean deleteTask(@NonNull String mapName, boolean isDownloadTask) {
+        Realm realm = Realm.getDefaultInstance();
+        MapBoxOfflineQueueTask taskToDelete;
+        if (isDownloadTask) {
```
Comment (pld) [inline library/src/main/java/io/ona/kujaku/data/realm/RealmDatabase.java]: how about removing the repetition by only setting the `taskType` in the conditional:

```java
MapBoxOfflineQueueTask taskType = isDownloadTask ? MapBoxOfflineQueueTask.TASK_TYPE_DOWNLOAD : MapBoxOfflineQueueTask.TASK_TYPE_DELETE

taskToDelete = realm.where(MapBoxOfflineQueueTask.class)
                    .equalTo("taskType", taskType)
                    .contains("task", mapName)
                    .findFirst();
```

**Comment by pld** [comment]
this also looks like it's missing unit tests

**Comment by ekigamba** [comment]
Yea. Most of the unit tests for this class are on another branch

File: library/src/main/java/io/ona/kujaku/data/realm/RealmDatabase.java (hunk 1 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/data/realm/RealmDatabase.java]: Btw, yea

**Comment by pld** [comment]
Including for this new method?

**Comment by ekigamba** [comment]
Nope, thought I would add that once the other branch was completed and merged. But I could also write other tests on this branch & documentation

**Comment by ekigamba** [comment]
sdfsdf

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 2]
```diff
@@ -359,15 +474,57 @@ private void getTaskStatus(MapBoxOfflineQueueTask mapBoxOfflineQueueTask, String
...
                     .setSmallIcon(R.drawable.ic_stat_file_download);
+
+            if (showAction) {
+                stopDownloadIntent = new Intent(this, MapboxOfflineDownloaderService.class);
+                stopDownloadIntent.putExtra(Constants.PARCELABLE_KEY_SERVICE_ACTION, SERVICE_ACTION.STOP_CURRENT_DOWNLOAD);
+                stopDownloadIntent.putExtra(Constants.PARCELABLE_KEY_MAP_UNIQUE_NAME, mapName);
+                stopDownloadIntent.putExtra(Constants.PARCELABLE_KEY_MAPBOX_ACCESS_TOKEN, mapBoxAccessToken);
+                stopDownloadIntent.putExtra(Constants.PARCELABLE_KEY_DELETE_TASK_TYPE, MapBoxOfflineQueueTask.TASK_TYPE_DOWNLOAD);
+
+                PendingIntent stopDownloadPendingIntent = PendingIntent.getService(this, 1, stopDownloadIntent, PendingIntent.FLAG_UPDATE_CURRENT);
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Please put the request code used in this pending intent in a constant.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 3]
```diff
@@ -359,15 +474,57 @@ private void getTaskStatus(MapBoxOfflineQueueTask mapBoxOfflineQueueTask, String
...
+                progressNotificationBuilder.addAction(stopDownloadAction);
+            }
+        }
+
+        if (percentageProgress == 0 && showAction) {
+            progressNotificationBuilder.setContentTitle("Offline Map Download Progress: " + mapName);
+
+            stopDownloadIntent.putExtra(Constants.PARCELABLE_KEY_MAP_UNIQUE_NAME, mapName);
+
+            PendingIntent stopDownloadPendingIntent = PendingIntent.getService(this, 1, stopDownloadIntent, PendingIntent.FLAG_CANCEL_CURRENT);
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Please put the request code used in this pending intent in a constant.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 4]
```diff
@@ -359,15 +474,57 @@ private void getTaskStatus(MapBoxOfflineQueueTask mapBoxOfflineQueueTask, String
      * @param mapName the unique map name
      * @param percentageProgress Download progress usually between 0-100%
      */
-    private void showProgressNotification(@NonNull String mapName, double percentageProgress) {
+    private void showProgressNotification(@NonNull String mapName, double percentageProgress, boolean showAction) {
         if (progressNotificationBuilder == null) {
             progressNotificationBuilder = new NotificationCompat.Builder(MapboxOfflineDownloaderService.this)
-                    .setContentTitle("Offline Map Download Progress: " + currentMapDownloadName)
+                    .setContentTitle("Offline Map Download Progress: " + mapName)
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Move to String resources XML please.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 5]
```diff
@@ -359,15 +474,57 @@ private void getTaskStatus(MapBoxOfflineQueueTask mapBoxOfflineQueueTask, String
...
+                PendingIntent stopDownloadPendingIntent = PendingIntent.getService(this, 1, stopDownloadIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+                NotificationCompat.Action stopDownloadAction = new NotificationCompat.Action(R.drawable.ic_mapbox_download_stop, getString(R.string.stop_download), stopDownloadPendingIntent);
+
+                progressNotificationBuilder.mActions.clear();
+                progressNotificationBuilder.addAction(stopDownloadAction);
+            }
+        }
+
+        if (percentageProgress == 0 && showAction) {
+            progressNotificationBuilder.setContentTitle("Offline Map Download Progress: " + mapName);
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Move to String resources XML please.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 6]
```diff
@@ -359,15 +474,57 @@ private void getTaskStatus(MapBoxOfflineQueueTask mapBoxOfflineQueueTask, String
...
+            progressNotificationBuilder.mActions.clear();
+            progressNotificationBuilder.addAction(stopDownloadAction);
+        }
+
+        // Remove all previous actions if showAction is false
+        if (!showAction) {
+            progressNotificationBuilder.mActions.clear();
         }
 
         progressNotificationBuilder.setContentText("Downloading: " + formatDecimal(percentageProgress) + " %");
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Move to String resources XML please.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 7]
```diff
@@ -470,16 +627,18 @@ public void onStatusChanged(OfflineRegionStatus status, OfflineRegion offlineReg
         sendBroadcast(SERVICE_ACTION_RESULT.SUCCESSFUL, currentMapDownloadName, currentServiceAction, String.valueOf(percentageDownload));
 
         if (status.isComplete()) {
+            stopDownloadProgressUpdater();
             showDownloadCompleteNotification("Download for " + currentMapDownloadName + " Map Complete!", "Downloaded " + getFriendlyFileSize(status.getCompletedResourceSize()) );
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Move to String resources XML please.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 8]
```diff
@@ -470,16 +627,18 @@ public void onStatusChanged(OfflineRegionStatus status, OfflineRegion offlineReg
...
-            showProgressNotification(currentMapDownloadName, percentageDownload);
+            //showProgressNotification(currentMapDownloadName, percentageDownload);
+            queueDownloadProgressUpdate(currentMapDownloadName, percentageDownload);
         }
     }
 
     @Override
-    public void onError(@NonNull String reason,@Nullable String message) {
+    public void onError(@NonNull String reason, @Nullable String message) {
         String finalMessage = "REASON : " + reason;
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Move to String resources XML please.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 9]
```diff
@@ -151,14 +188,62 @@ private boolean persistOfflineMapTask(@Nullable Intent intent) {
...
+                                            .getInstance(MapboxOfflineDownloaderService.this, mapBoxAccessToken)
+                                            .deleteMap(mapUniqueName, new OfflineRegion.OfflineRegionDeleteCallback() {
+
+                                                @Override
+                                                public void onDelete() {
+                                                    if (deleteTaskFromRealmDatabase(taskType, mapUniqueName)) {
+                                                        sendBroadcast(SERVICE_ACTION_RESULT.SUCCESSFUL, mapUniqueName, SERVICE_ACTION.STOP_CURRENT_DOWNLOAD);
+
+                                                        showDownloadCompleteNotification("Download for " + mapUniqueName + " stopped!", "All downloaded resources for the map have been deleted also");
+                                                        performNextTask();
```
Comment (jasonrogena) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Looks like `performNextTask()` should be executed regardless of the condition. Could you please move it to after the if-else blocks.

**Comment by jasonrogena** [comment]
@ekigamba Just a suggestion. As a way of decongesting the `MapboxOfflineDownloaderService` is it possible to move the notification logic to a separate class. Make it reusable for other services/activities that might want to use notifications.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java (hunk 9 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Yea

**Pull request merged**